content/blog/2025/fundamentals_architecture.md

```diff
@@ -0,0 +1,236 @@
+
+This fundamental contrast is why software architecture must be flexible and modular, allowing for ongoing adjustments without needing a complete rebuild — something rarely possible in physical construction.
+
+## Story of one "software-like construction."
+
+One of the key skills of a software architect is the ability to explain complex concepts using simple, real-life examples. Let me illustrate how software architecture and development can be visualized and explained through construction terms.
+
+### Beginning
+
+ABC company decides to construct a building to rent out commercial spaces. After conducting market research, they identify a suitable location. However, at this early stage, it is difficult to predict exactly how many tenants will occupy the space or what types of businesses will operate there.
+
+To move forward, an architect must make some reasonable assumptions to create a practical design:
+
+* The most common use for commercial spaces will be offices, so the design should primarily accommodate office layouts.
+* The average company size in the area is about **100 employees**, so spaces should be designed to fit businesses of this scale.
+* To maximize flexibility, the building will have **two office spaces per floor**, allowing for different tenant needs.
+* As an initial plan, construction will start with **five floors**, with the potential to expand if demand grows.
+
+The business team reviews these parameters, finds them reasonable, and gives approval. With these assumptions agreed upon, construction begins.
+
+{{< image src="../../../images/blog/2025/fundamentals_architecture/stage-1.drawio.png" alt="Stage 1" >}}
+
+### So far so good
+
+Six months later, **three out of five floors** are built and fully operational. The first tenants are already moving in, and negotiations with potential new tenants are ongoing. Seeing the strong demand, the business decides to **increase the building’s height to 10 floors**.
+
+While this expansion wasn’t originally planned, the architect confirms that the foundation is strong enough to support up to 10 floors. However, he warns the business that **12 floors is the absolute maximum** the current foundation can handle.
+
+If the company anticipates needing more space beyond that, it should consider starting a new building immediately to avoid future bottlenecks. However, due to budget constraints, the business has decided to stick with 10 floors and revisit expansion plans if demand continues to grow.
+
+{{< image src="../../../images/blog/2025/fundamentals_architecture/stage-2.drawio.png" alt="Stage 2" >}}
+
+### We have to win this customer
+
+With **six out of ten floors** already built, a sales representative rushes into the architect's office, holding a stack of papers.
```

Comment:
It has to be the building site to be accurate XD

Reply:
Discussed on call yesterday. We are assuming "positive" (properly established processes) scenario here, so resolving.